Question 1(a)

In [21]:
import pandas as pd
student=pd.read_csv('Students_Performance.csv')
gender_count=student['gender'].value_counts() #'gender_count' will store count of male aand female candidates as a series
print('{0} number of males and {1} number of females participted in the test'.format(gender_count['male'],gender_count['female']))

482 number of males and 518 number of females participted in the test


Question 1(b)

In [23]:
student['parental level of education'].value_counts()

parental level of education
some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: count, dtype: int64

Conclusion from above output are as follows:
From the above output, it can be seen that parents have six types of education levels:
375 parents have completed only up to high school level, 
226 have attended college education, 
399 have acquired some king of degree (222 associates, 118 bachelors and 59 masters)

Question 1(c) - (i)  #Average Performance based on Gender

In [9]:
#This pgm finds students of which gender (male or female) performs best on each subject based on average score.
#Avg marks obtained by males and females is calculated separately for each subject, viz, math, reading and writing.
#Then, avg of male and female of each subject are compared and results are printed.
#'\033[1;4m makes the text bold and underlined
print('\033[1;4m(i)Average Performance based on Gender\n\033[0m')
#calculate avg of math, reading and writing based on gender and results are stored in table 'avgScore' 
avgScore=student.groupby('gender').agg({'math score':['mean'],'reading score':['mean'],'writing score':['mean']}) 

#this for loop iterates through the table 'avgScore' column wise. Avg of math score of male and female is taken in 1st iteration,
#Avg of reading score of male and female is taken in 2nd iteration, Avg of writing score of male and female is taken in 3rd iteration.
#In each iteration, avg marks (in one subject) of male and female are compared and results displayed.
for i in avgScore.keys(): 
    score_female,score_male=avgScore[i].iloc[0],avgScore[i].iloc[1] #store avg marks from table 'avgScore' into two variables
    
    if(score_female>score_male):
        print("Female students got the most on average for", i[0])
    else:
        print("Male students got the most on average for", i[0])

avgScore  #display the calculated avg of math, reading and writing based on gender

(i)Average Performance based on Gender

Male students got the most on average for math score
Female students got the most on average for reading score
Female students got the most on average for writing score


math score reading score writing score
             mean          mean          mean
gender                                       
female  63.376448     72.590734     72.467181
male    68.821577     65.545643     63.446058

Question 1(c) - (ii) #Average Performance based on Test Preparation Course

In [11]:
#this pgm finds whether the students who competed test preparation course performs best on each subject based on average score.
print('\033[1;4m\n(ii)Average Performance based on Test Preparation Course\n\033[0m')
avgScore=student.groupby('test preparation course').agg({'math score':['mean'],'reading score':['mean'],'writing score':['mean']})

for i in avgScore.keys():
    score_completed,score_none=avgScore[i].iloc[0],avgScore[i].iloc[1]

    if(score_completed>score_none):
        print("Students who completed test preparation course got the most on average for", i[0])
    else:
        print("Students who do not completed test preparation course got the most on average for", i[0])

avgScore


(ii)Average Performance based on Test Preparation Course

Students who completed test preparation course got the most on average for math score
Students who completed test preparation course got the most on average for reading score
Students who completed test preparation course got the most on average for writing score


math score reading score writing score
                              mean          mean          mean
test preparation course                                       
completed                 69.96648     74.175978     74.684358
none                      63.78972     66.417445     64.457944

Question 1(d)

In [288]:
#Variation in scores is taken in the form of standard deviation
print('\033[1;4mFollowing table shows the scoring variation for math, reading and writing based on Gender:\n\033[0m')
std_Score_gender=student.groupby('gender').agg({'math score':['std','max','min'],'reading score':['std','max','min'],'writing score':['std','max','min']})
std_Score_gender

Following table shows the scoring variation for math, reading and writing based on Gender:



math score          reading score          writing score         
              std  max min           std  max min           std  max min
gender                                                                  
female  16.029928  100   0     14.411018  100  17     14.844842  100  10
male    14.556411  100  27     14.149594  100  23     14.227225  100  15

In [274]:
print('\033[1;4m\nFollowing table shows the scoring variation for math, reading and writing based on Test preparation Course:\n\033[0m')
std_Score_preparation=student.groupby('test preparation course').agg({'math score':['std','max','min'],'reading score':['std','max','min'],'writing score':['std','max','min']})
std_Score_preparation


Following table shows the scoring variation for math, reading and writing based on Test preparation Course:



math score          reading score           \
                               std  max min           std  max min   
test preparation course                                              
completed                14.521847  100  23     13.537572  100  37   
none                     15.705689  100   0     14.608896  100  17   

                        writing score           
                                  std  max min  
test preparation course                         
completed                   13.236412  100  36  
none                        15.041667  100  10

From above output, it is seen that marks scored by both males and females are showing equal variance/std deviation from mean.
ie, spread of each students marks from corresponding mean will be approximately same. Min and max marks scored in each subject gender wise and preparation strategy wise also displayed. Max marks is same for al categoreis, but minimum marks in each subject show some variation for all genders and test preparation course.

Question 1(e)

In [15]:
student['math score'].quantile(q=0.75) #find the math score which comes at 25 the percentile
pd.set_option('display.max_rows',None) #to set displaying all columns in pandas output. By default, only first and last 5 rows will be displayed.
#25th percentile value of math score was 77. So, decided to give bonus points to all students with math score > = 77 
student[student['math score']>=77].sort_values(by=['math score'],axis=0,ascending=False)

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
998    male        group E          associate's degree  free/reduced   
997    male        group D                some college      standard   
996    male        group A                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
515  female        group E                some college      standard   
995    male        group E                some college      standard   
514  female        group E                 high school      standard   
513  female        group E           bachelor's degree      standard   
994    male        group C          associate's degree      standard   
993    male        group C                some college      standard   
512  female        group D                some college      standard   
510  female        group B           bachelor's degree      standard   
511  female        group D            some high school      standard   
992    male        group E          associate's degree      standard   
989    male        group A          associate's degree      standard   
990    male        group C          associate's degree      standard   
991    male        group E                some college      standard   
508  female        group C           bachelor's degree      standard   
509  female        group C          associate's degree      standard   
988    male        group C           bachelor's degree      standard   
507  female        group E          associate's degree      standard   
987    male        group D             master's degree      standard   
985    male        group E          associate's degree      standard   
505  female        group B          associate's degree      standard   
506  female        group E             master's degree      standard   
986    male        group E                 high school      standard   
984    male        group E            some high school      standard   
983    male        group C           bachelor's degree      standard   
982    male        group B            some high school      standard   
503  female        group D           bachelor's degree  free/reduced   
504  female        group E          associate's degree      standard   
981    male        group E                some college  free/reduced   
980    male        group C                some college      standard   
978    male        group C          associate's degree      standard   
979    male        group E            some high school      standard   
499  female        group A            some high school      standard   
500  female        group C           bachelor's degree      standard   
501  female        group D             master's degree      standard   
502  female        group E           bachelor's degree  free/reduced   
497  female        group C          associate's degree      standard   
498  female        group C          associate's degree      standard   
971    male        group A           bachelor's degree      standard   
977    male        group E          associate's degree  free/reduced   
976    male        group C             master's degree      standard   
975    male        group C           bachelor's degree      standard   
974    male        group C                some college      standard   
973    male        group B                some college      standard   
972    male        group B          associate's degree      standard   
964    male        group B          associate's degree      standard   
495  female        group B             master's degree      standard   
969    male        group E             master's degree      standard   
968    male        group D          associate's degree      standard   
967    male        group D          associate

Question 2(a)

In [17]:
import pandas as pd
import numpy as np
import scipy.stats as stats
from statsmodels.stats.weightstats import ztest

sales=pd.read_csv('Sales_add.csv')
pop_mean=sales['Sales_before_digital_add(in $)'].mean() #before sales avg
sample_mean=sales['Sales_After_digital_add(in $)'].mean()#after sales avg
sample_std_dev=sales['Sales_After_digital_add(in $)'].std() #since t test is used, sample std deviation taken in calculation
sample_size=sales['Month'].size

#Null hyp H0: There is no improvemnet in sales
#Alt hyp Ha: There is improvemnet in sales after digital marketing
#t-test chosen because sample size<30
alpha=0.05 #confidence interval taken as 95%

data=np.array(sales['Sales_After_digital_add(in $)'])
t_stats,p_value=stats.ttest_1samp(data,pop_mean,alternative="greater") #since we want to prove there is an increase in sales, one sided t test is done.
tcritical=1.721 # from one sided t table
print(t_stats,p_value)

if t_stats > tcritical:
  print("Reject null hyp: There is improvemnet in sales after digital marketing")
else:
  print("fail to reject null hyp: There is no improvemnet in sales")

15.028066221386988 5.189045779399668e-13
Reject null hyp: There is improvemnet in sales after digital marketing


Question 2(b)

In [19]:
import numpy as np
import scipy.stats as stats

#Null hyp H0: there is no dependency between the features Region and Manager
#Alt hyp Ha: there is dependency between the features Region and Manager
observed=np.array([[4,4,1],[3,1,3],[3,2,1]]) #first array [4,4,1] represents number of manager A region wise, 
                                             #second array [3,1,3] represents number of manager B region wise,
                                             #third array [3,2,1] represents number of manager C region wise
chi2_stat,p_value,dof,expected = stats.chi2_contingency(observed)
print(chi2_stat,p_value,dof,expected)
dof=4 #(no of rows-1)=2, (no of col-1)=2
alpha=0.05 #confidence interval taken as 95%
chi_critical=9.488 #from chi square table corresponding to dof=4 and alpha =0.05
print('\nThe result is:')
if p_value < alpha:
  print("Reject null hyp: there is dependency between the features Region and Manager.")
else:
  print("Fail to reject null hyp: there is no dependency between the features Region and Manager.")

3.050566893424036 0.5493991051158094 4 [[4.09090909 2.86363636 2.04545455]
 [3.18181818 2.22727273 1.59090909]
 [2.72727273 1.90909091 1.36363636]]

The result is:
Fail to reject null hyp: there is no dependency between the features Region and Manager.
